posix/mount.c

```diff
@@ -0,0 +1,286 @@
+    const char* filesystemtype,
+    unsigned long mountflags,
+    const void* data)
+{
+    int ret = -1;
+    oe_device_t* device = NULL;
+    oe_device_t* new_device = NULL;
+    bool locked = false;
+
+    OE_UNUSED(data);
+
+    if (!target || !filesystemtype)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    /* Resolve the device for the given filesystemtype. */
+    if (!(device =
+              oe_device_table_find(filesystemtype, OE_DEVICE_TYPE_FILE_SYSTEM)))
+        OE_RAISE_ERRNO_MSG(OE_EINVAL, "filesystemtype=%s", filesystemtype);
+
+    /* Be sure the full_target directory exists (if not root). */
+    if (oe_strcmp(target, "/") != 0)
+    {
+        struct oe_stat buf;
+        int retval = -1;
+
+        if ((retval = oe_stat(target, &buf)) != 0)
+            OE_RAISE_ERRNO(oe_errno);
+
+        if (!OE_S_ISDIR(buf.st_mode))
+            OE_RAISE_ERRNO(OE_ENOTDIR);
+    }
+
+    /* Lock the mount table. */
+    oe_conditional_lock(&_lock, &locked);
+
+    /* Install _free_mount_table() if not already installed. */
+    if (_installed_free_mount_table == false)
+    {
+        oe_atexit(_free_mount_table);
+        _installed_free_mount_table = true;
+    }
+
+    /* Fail if mount table exhausted. */
+    if (_mount_table_size == MAX_MOUNT_TABLE_SIZE)
+        OE_RAISE_ERRNO(OE_ENOMEM);
+
+    /* Reject duplicate mount paths. */
+    for (size_t i = 0; i < _mount_table_size; i++)
+    {
+        if (oe_strcmp(_mount_table[i].path, target) == 0)
+            OE_RAISE_ERRNO(OE_EEXIST);
+    }
+
+    /* Clone the device. */
+    if (device->ops.fs.clone(device, &new_device) != 0)
+        OE_RAISE_ERRNO(oe_errno);
+
+    /* Assign and initialize new mount point. */
+    {
+        size_t index = _mount_table_size;
+        size_t len = oe_strlen(target);
+
+        _mount_table[index].path_size = len + 1;
+        _mount_table[index].path = oe_malloc(len + 1);
+
+        if (!_mount_table[index].path)
+            OE_RAISE_ERRNO(OE_ENOMEM);
+
+        if (oe_memcpy_s(
+                _mount_table[index].path,
+                _mount_table[index].path_size,
+                target,
+                len + 1) != OE_OK)
+        {
+            OE_RAISE_ERRNO(OE_EINVAL);
+        }
+
+        _mount_table[index].fs = new_device;
+        _mount_table_size++;
+    }
+
+    /* Notify the device that it has been mounted. */
+    if (new_device->ops.fs.mount(new_device, source, target, mountflags) != 0)
+    {
+        oe_free(_mount_table[--_mount_table_size].path);
+        goto done;
+    }
+
+    new_device = NULL;
+    ret = 0;
+
+done:
+
+    oe_conditional_unlock(&_lock, &locked);
+
+    if (new_device)
+        new_device->ops.device.release(new_device);
+
+    return ret;
+}
+
+int oe_umount2(const char* target, int flags)
+{
+    int ret = -1;
+    size_t index = (size_t)-1;
+    char suffix[OE_PATH_MAX];
+    oe_device_t* device = oe_mount_resolve(target, suffix);
+    bool locked = false;
+
+    OE_UNUSED(flags);
```

Comment:
>OE_UNUSED(flags); [](start = 4, length = 17)

This will be surprising to a developer if they invoke `umount2` explicitly with non-zero flags. We should check for that and return ENOTSUP.

Reply:
Passing the flags onto the device. Changed device function name to **umount2** and added flags parameter.